api/provider/aws/logging.go

```diff
@@ -0,0 +1,165 @@
+package aws
+
+import (
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs/cloudwatchlogsiface"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+const (
+	// DescribeLogStreams is throttled after five transactions per second.
+	// With 50 streams/transaction, 1000 gives a reasonable streams:time ratio
+	MAX_DESCRIBE_STREAMS_COUNT = 1000
+)
+
+func GetLogsFromTaskARNs(
+	cloudWatchLogsAPI cloudwatchlogsiface.CloudWatchLogsAPI,
+	logGroupName string,
+	taskARNs []string,
+	tail int,
+	start,
+	end time.Time,
+) ([]models.LogFile, error) {
+	logFiles := []models.LogFile{}
+	fn := func(containerName string, events []*cloudwatchlogs.OutputLogEvent) {
+		logFile := models.LogFile{
+			ContainerName: containerName,
+		}
+
+		logFile.Lines = make([]string, len(events))
+		for i, event := range events {
+			logFile.Lines[i] = aws.StringValue(event.Message)
+		}
+
+		logFiles = append(logFiles, logFile)
+	}
+
+	if err := iterateLogEvents(
+		cloudWatchLogsAPI,
+		logGroupName,
+		taskARNs,
+		tail,
+		start,
+		end,
+		fn); err != nil {
+		return nil, err
+	}
+
+	return logFiles, nil
+}
+
+func iterateLogEvents(
+	cloudWatchLogsAPI cloudwatchlogsiface.CloudWatchLogsAPI,
+	logGroupName string,
+	taskARNs []string,
+	tail int,
+	start time.Time,
+	end time.Time,
+	fn func(logStreamName string, events []*cloudwatchlogs.OutputLogEvent),
+) error {
+	taskIDMatch := generateTaskIDMap(taskARNs)
+
+	logStreams, err := describeLogStreams(cloudWatchLogsAPI, logGroupName)
```

Comment:
[nit] I don't want to be too dogmatic about the patterns we've adopted in the `provider` layer, but it seems like we could do something similar here in regards to how we call helper functions and there would be some benefit to doing so. For example, `GetLogsFromTaskARNs()` could call the various helpers rather than having the helpers call each other --- I've found this to be a lot more readable.